prevent fetching spellcheck Dictionary

gatzjames opened the pull request:
Disables spellchecking in electron and sets the dictionary download URL to `https://00.00/` as a way to prevent fetching anything from Chromium's CDN.


TODO:
- [x] Validate the behaviour in Linux/Windows

Closes INS-891

gatzjames committed:
Disable spellchecking and prevent fething translations from Chromium's CDN

dimitropoulos commented on packages/insomnia-app/app/main/window-utils.ts:
```diff
@@ -20,11 +20,20 @@ import { clickLink, getDataDirectory, restartApp } from '../common/electron-help
 app.allowRendererProcessReuse = false;
 
+app.on('ready', () => {
+  // There's no option that prevents Electron from fetching spellcheck dictionaries from
+  // Chromium's CDN and passing a non-resolving URL is the only known way to prevent it from
+  // fetching.  https://github.com/electron/electron/issues/22995
```
nit: no reason to break lines unnecessarily:
![Screenshot_20210914_202710](https://user-images.githubusercontent.com/15232461/133351120-e97e750f-4af5-4b2d-ae4f-1488d7123bf5.png)

```suggestion
  // There's no option that prevents Electron from fetching spellcheck dictionaries from Chromium's CDN and passing a non-resolving URL is the only known way to prevent it from fetching.
  // see: https://github.com/electron/electron/issues/22995
```

develohpanda commented on packages/insomnia-app/app/main/window-utils.ts:
```diff
@@ -20,11 +20,20 @@ import { clickLink, getDataDirectory, restartApp } from '../common/electron-help
+  // fetching.  https://github.com/electron/electron/issues/22995
+  // On macOS the OS spellchecker is used and therefore we do not download any dictionary files.
+  // This API is a no-op on macOS.
+  session.defaultSession.setSpellCheckerDictionaryDownloadURL('https://00.00/');
```
Would setting it this way also apply the same setting to OAuth windows that open, or just to the main window of the app?

What is the best way to validate this - through WireShark for instance, or were you able to detect it though electron itself?

dimitropoulos replied:
we finally figured out an approach that was reliable: check out https://github.com/Kong/insomnia/pull/4030#pullrequestreview-756340122

develohpanda commented on packages/insomnia-app/app/main/window-utils.ts:
```diff
@@ -85,6 +94,7 @@ export function createWindow() {
     acceptFirstMouse: true,
     icon: path.resolve(__dirname, appLogo),
     webPreferences: {
+      spellcheck: false,
```
Worth double checking where else we create new windows, there are a few, here's one for example:
https://github.com/Kong/insomnia/blob/ccac337c4f327329f32bb93ac1324cfe28562c2b/packages/insomnia-app/app/network/o-auth-2/misc.ts#L76-L83

dimitropoulos replied:
an update on this: @gatzjames, @johnwchadwick and I discussed this and determined that we actually want to leave it on.  Ultimately, the thing we are trying to do is prevent the fetch from happening.  But for MacOS users, there _is no_ fetch because chromium uses builtin functionality.  It's not like we want to turn off spellchecking as a product goal - we just want to disable the fetch.  If linux and windows both worked like mac, then we'd never be in this situation digging around.  So, based on that, there's nothing to lose from leaving this enabled.  If there's no dictionary then spellchecking will just not work, anyway, so there's no error or anything like that that will surface to the user.

gatzjames committed:
Update packages/insomnia-app/app/main/window-utils.ts

Co-authored-by: Dimitri Mitropoulos <dimitrimitropoulos@gmail.com>

dimitropoulos committed:
don't unnecessarily disable spellcheck

since, after all, it doesn't download in Mac and upcoming versions of Windows.  it's not like we have a thing against spellchecking - the ticket is to disable a dictionary download.

dimitropoulos reviewed (approved):
before this PR, when running 

```
strace -fe trace=network -o network2.log npm run app-start
```

The following two lines were observed.

```
162898 sendto(26, ";\35\1 \0\1\0\0\0\0\0\1\nredirector\4gvt1\3com"..., 48, MSG_NOSIGNAL, NULL, 0) = 48
162898 recvfrom(26, ";\35\201\200\0\1\0\1\0\0\0\1\nredirector\4gvt1\3com"..., 1024, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, [28->16]) = 64
```

After this PR, both lines were gone.

develohpanda commented:
@dimitropoulos I feel like that's a valuable note to make in our Confluence space so that if/when we ever need to investigate again we know what to do!
